scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordered_serialization/providers/StringOrderedBuf.scala

```diff
@@ -0,0 +1,122 @@
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+package com.twitter.scalding.macros.impl.ordered_serialization.providers
+
+import scala.language.experimental.macros
+import scala.reflect.macros.Context
+
+import com.twitter.scalding._
+import com.twitter.scalding.macros.impl.ordered_serialization.{ CompileTimeLengthTypes, ProductLike, TreeOrderedBuf }
+import CompileTimeLengthTypes._
+import java.nio.ByteBuffer
+import com.twitter.scalding.serialization.OrderedSerialization
+
+object StringOrderedBuf {
+  def dispatch(c: Context): PartialFunction[c.Type, TreeOrderedBuf[c.type]] = {
+    case tpe if tpe =:= c.universe.typeOf[String] => StringOrderedBuf(c)(tpe)
+  }
+
+  def apply(c: Context)(outerType: c.Type): TreeOrderedBuf[c.type] = {
+    import c.universe._
+
+    def freshT(id: String) = newTermName(c.fresh(id))
+
+    new TreeOrderedBuf[c.type] {
+      override val ctx: c.type = c
+      override val tpe = outerType
+      override def compareBinary(inputStreamA: ctx.TermName, inputStreamB: ctx.TermName) = {
+        val lenA = freshT("lenA")
+        val lenB = freshT("lenB")
+
+        q"""
+        val $lenA = $inputStreamA.readPosVarInt
+        val $lenB = $inputStreamB.readPosVarInt
+        _root_.com.twitter.scalding.serialization.StringOrderedSerialization.binaryIntCompare($lenA,
+          $inputStreamA,
+          $lenB,
+          $inputStreamB)
+      """
+      }
+
+      override def hash(element: ctx.TermName): ctx.Tree = q"_root_.com.twitter.scalding.serialization.Hasher.string.hash($element)"
+
+      override def put(inputStream: ctx.TermName, element: ctx.TermName) = {
+        val bytes = freshT("bytes")
+        val charLen = freshT("charLen")
+        val len = freshT("len")
+        q"""
+         // Ascii is very common, so if the string is short,
+         // we check if it is ascii:
+         def isShortAscii(size: Int, str: String): Boolean = (size < 65) && {
+           var pos = 0
+           var ascii: Boolean = true
+           while((pos < size) && ascii) {
+             ascii = (str.charAt(pos) < 128)
+             pos += 1
+           }
+           ascii
+         }
+
+         val $charLen = $element.length
+         if ($charLen == 0) {
+           $inputStream.writePosVarInt(0)
+         }
+         else if (isShortAscii($charLen, $element)) {
+           $inputStream.writePosVarInt($charLen)
+           val $bytes = new Array[Byte]($charLen)
+           // This deprecated gets ascii bytes out, but is incorrect
+           // for non-ascii data.
+           $element.getBytes(0, $charLen, $bytes, 0)
+           $inputStream.write($bytes)
+         }
+         else {
+           // Just use utf-8
+           val $bytes = $element.getBytes("UTF-8")
```

Comment:
maybe add a comment here along the lines of
`// TODO: Investigate benefit of using getBytes(Charset) on jvms > 1.6`